### ISSUE TITLE
Problem: ZendFramework 1 is incompatible with Php 7.2 and thus Ubuntu 18.04

### ISSUE THREAD

**Issue by Robbt** [LibreTime/libretime#580]
So, Php 7.2 introduced some inheritance tightening that causes ZF1 to break. See this [post](https://stackoverflow.com/questions/49072332/is-zend-framework-1-compatible-with-php-7-2). 

Ubuntu 18.04 uses Php 7.2 by default and so LT is no longer going to work with this without some workarounds.

One option would be to with the installer force Ubuntu 18.04 to use an older version of PHP, I think Php 7.1 should still work.

Another option would be to switch to a community fork of Zend 1 Framework that includes Php 7.2 fixes. But I don't think there really is one here is a [discussion](https://discourse.zendframework.com/t/zf1-community-maintained-official-fork/419/5) that doesn't end in an actual project and includes some reasons why this isn't a good idea. Someone did make individual [composer packages](https://github.com/zf1s) but it doesn't appear to be very well maintained and there doesn't seem to be a fix in zend_rest for the issue we are running into.

It should work on Php 7.0 which is supported until 3 Dec 2018. and PHP 7.1 is supported until 1 Dec 2019. The breakages all happen with Php 7.2

**Comment by hairmare** [comment]
*sigh* 7.2 is going to be relevant really soon. We might be able to get away by switching to the https://github.com/magento/zf1 fork that seems to support 7.2 as far as I can tell. I haven't tried yet but i did some digging through available zf1 forks.

**Comment by Robbt** [comment]
Yeah none of the forks looked promising from the bit of digging I did. They all rely upon an unsupported version of PHPUnit and have no tests for anything beyond Php 7.0. Also none of them appeared to modify the Rest server php file that was erroring when I tried the bionic vagrant box. There doesn't appear to be a real supported fork, so we might have to start work on migrating sooner than we like and just set 3.0 to work on older PHP versions and distros that support it. The discussion I linked to above was pretty relevant because it featured the people who had been maintaining Zend 1 before it was EOL and their perspective on the work required to keep it running with new versions of PHP.

**Comment by hairmare** [comment]
I haven't really done a deep dive into this (yet) but I might have some reasons to do so for non-LibreTime related projects... For now it looks like we won't be supporting 7.2 in the foreseeable short term future.

**Comment by Robbt** [comment]
I did some basic testing on 7.1 on a vagrant virtual box Ubuntu 16.04 and it didn't break anything that I can find. So in theory this gives us another year worth of security updates since Php 7.0 gets its last security updates at the end of this year but 7.1 will have them until 12/1/2019.

**Comment by hairmare** [comment]
For me the upcoming CentOS 8 release will probably have me looking into this. The recently released RHEL 8 beta has php 7.2 and will most likely stay on that version for its life cycle. I've yet to look into the beta to figure out if there will be a 7.1 php application stream and how it will be life cycled.

**Comment by Robbt** [comment]
So I see two ways forward to support 18.04 - one being to simply use a PPA with backported PHP 7.1 - this makes whomever installs it dependent upon the person running the PPA forever. If that person quits updating it or removes it that will break or leave their software vulnerable. This is better than having a non-working install but it isn't ideal.

The other option would be for us to contribute to a fork of zf1 to workaround the various incompatibilities introduced by php 7.2. This is also somewhat sketchy because we don't know how much work it will entail and it would leave us maintaining a forked version of ZF1. This might be a good idea if we could build upon the work done by someone else but it is opening up a new can of worms in terms of potential bugs. It would probably make sense at this point to also devendorize the zend 1 framework and reinclude it inside of libretime since we are basically taking ownership over any changes.

To test this we could try the two active forks that I'm aware of - the magneto one @hairmare references and the one with individual packages. So that is a lot more work in the long run but could help us stopgap keep things running as they are with supported PHP environment as we plan for the longterm #2 upgrade. If I feel like it I might do further testing on the community forks because migrating away is also a huge job that we haven't come up with a plan for.

**Comment by paddatrapper** [comment]
To add to this - Debian Buster will only support PHP 7.3...

**Comment by hairmare** [comment]
https://github.com/magento/zf1 is out since they removed Zend/Application.php in their 1.14.x releases when they added php 7.2 support. I haven't given the zf1s packages a shot yet but they do [seem to have the file](https://github.com/zf1s/zend-application).

**Comment by hairmare** [comment]
I got pretty far with the zf1s forks but some php 7.2 issues still remain. I'm fixing the first one I found and trying to figure out if contributing to the zf1s repos makes sense in https://github.com/zf1s/zend-rest/pull/1. 

You can find my zf1 based branch at https://github.com/radiorabe/libretime/pull/new/dev/zf1s.

**Comment by Robbt** [comment]
Very cool. If we can get this working it will give us some breathing room to figure out the best upgrade path for future development.

**Comment by hairmare** [comment]
I also found https://github.com/Shardj/zf1 which does fix some php7.2 issues. I'll give it a try if we figure out the zf1s orga isn't maintained/doesn't accepts patches. In any case it will also need some changes to support our critical path.

**Comment by hairmare** [comment]
These are next on the list to test but none of them support Zend_Rest as far as I can tell.

* https://github.com/cakeweb/zf1
* https://github.com/BassRefleXive/zf1

I looks like most forks end up removing the parts they don't need. It looks like everyone seems to drop support for `Zend_Application` at some point in their forks.

I looked at the following forks that won't work due to this.
* https://github.com/Icinga/icingaweb2-vendor-zf1
* https://github.com/diablomedia/zf1

**Comment by hairmare** [comment]
On my zf1s branch https://github.com/LibreTime/libretime/compare/master...radiorabe:dev/zf1s basic playout is now working, I did some testing and will also do some retesting on trusty to be sure I don't break anything old. I'll open a PR a soon as I get trusty tested.

It's already quite clear that I'll need some help in testing the PR as it's rather likely that I didn't catch everything in my tests. The critical path for my station (the program grid, basic uploading and, playout) as well as the podcast downloading work and I haven't found any pages in the UI that still throw 500 errors.

**Comment by frecuencialibre** [comment]
awesome! i'll do my best to test asap in january. should testing happen in php 7.2 or 7.3 for buster et al?

**Comment by hairmare** [comment]
I haven't tested 7.3 yet but AFAICT there shouldn't be any additional issues that we didn't fix for 7.2. If you would like to test 7.3 feel free to post any backtraces from zendphp.log or the apache error.log and I'll try to get them fixed as well.

I'll also take some time to re-test everything on trusty to ensure we still support that. But I don't think there should be any issues there.

**Comment by Robbt** [comment]
Excited to start testing this. I like this option more than pushing people to use an outdated distro or version of PHP.

**Comment by Shardj** [comment]
Apparently I was mentioned here, I'll add that I have a working 7.2 branch of zf1 you can find here https://github.com/Shardj/zf1

**Comment by Robbt** [comment]
One thing I just realized after merging is we need instructions for how to upgrade. They will need to run composer-install after they upgrade to the latest version of Ubuntu - this should go into the release notes for sure.

**Comment by hairmare** [comment]
For people installing from rich tarballs no changes are needed. For other running the installer *should* run composer for them.
